rosbag2_transport/src/rosbag2_transport/qos.cpp

```diff
@@ -68,3 +73,126 @@ bool convert<rosbag2_transport::Rosbag2QoS>::decode(
+    request_qos.reliable();
+  } else {
+    if (reliability_reliable_endpoints_count > 0) {
+      ROSBAG2_TRANSPORT_LOG_WARN_STREAM(
+        "Some, but not all, publishers on topic \"" << topic_name << "\" "
+          "are offering RMW_QOS_POLICY_RELIABILITY_RELIABLE. "
+          "Falling back to RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT "
+          "as it will connect to all publishers. "
+          "Some messages from Reliable publishers could be dropped.");
+    }
+    request_qos.best_effort();
+  }
+
+  // Policy: durability
+  // If all publishers offer transient_local, we can request it and receive latched messages
+  if (durability_transient_local_endpoints_count == num_endpoints) {
+    request_qos.transient_local();
+  } else {
+    if (durability_transient_local_endpoints_count > 0) {
+      ROSBAG2_TRANSPORT_LOG_WARN_STREAM(
+        "Some, but not all, publishers on topic \"" << topic_name << "\" "
+          "are offering RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL. "
+          "Falling back to RMW_QOS_POLICY_DURABILITY_VOLATILE "
+          "as it will connect to all publishers. "
+          "Previously-published latched messages will not be retrieved.");
+    }
+    request_qos.durability_volatile();
+  }
+  // Policy: deadline
+  // Deadline does not affect delivery of messages,
+  // and we do not record Deadline"Missed events.
+  // We can always use unspecified deadline, which will be compatible with all publishers.
+
+  // Policy: lifespan
+  // Lifespan does not affect compatibiliy
+
+  // Policy: liveliness, liveliness_lease_duration
+  // Liveliness does not affect delivery of messages,
+  // and we do not record LivelinessChanged events.
+  // We can always use unspecified liveliness, which will be compatible with all publishers.
+  return request_qos;
+}
+
+namespace
+{
+bool operator==(const rmw_time_t & lhs, const rmw_time_t & rhs)
+{
+  return lhs.sec == rhs.sec && lhs.nsec == rhs.nsec;
+}
+
+bool all_profiles_effectively_same(const std::vector<Rosbag2QoS> & profiles)
+{
+  auto iterator = profiles.begin();
+  const auto p_ref = iterator->get_rmw_qos_profile();
+  iterator++;
+  for (; iterator != profiles.end(); iterator++) {
+    const auto p_next = iterator->get_rmw_qos_profile();
+    bool compatibility_equals_previous = (
+      // excluding history
+      p_ref.reliability == p_next.reliability &&
+      p_ref.durability == p_next.durability &&
+      p_ref.deadline == p_next.deadline &&
+      // excluding lifespan
+      p_ref.liveliness == p_next.liveliness &&
+      p_ref.liveliness_lease_duration == p_next.liveliness_lease_duration
+    );
+    if (!compatibility_equals_previous) {
+      return false;
+    }
+  }
+  return true;
+}
+}  // unnamed namespace
+
+Rosbag2QoS Rosbag2QoS::adapt_offer_to_recorded_offers(
+  const std::string & topic_name, const std::vector<Rosbag2QoS> & profiles)
+{
+  static const Rosbag2QoS default_offer{};
+  if (profiles.empty()) {
+    return default_offer;
```

Comment:
Since you're returning a copy, why not just do
```c++
return Rosbag2Qos{};
```
?

Reply:
I wanted to make it quite clear what it is that we're returning. In the `request` logic I use it twice.